Gems/PhysX/Code/Source/Scene/PhysXScene.cpp

```diff
@@ -974,27 +974,29 @@ namespace PhysX
         // Query flags.
         const physx::PxQueryFlags queryFlags = SceneQueryHelpers::GetPxQueryFlags(request->m_queryType);
         const physx::PxQueryFilterData queryData(queryFlags);
-
-        if (azrtti_istypeof<AzPhysics::RayCastRequest>(request))
-        {
-            return Internal::RayCast(azdynamic_cast<const AzPhysics::RayCastRequest*>(request),
-                s_rayCastBuffer, m_pxScene, queryData, m_raycastBufferSize);
-        }
-        else if (azrtti_istypeof<AzPhysics::ShapeCastRequest>(request))
-        {
-            return Internal::ShapeCast(azdynamic_cast<const AzPhysics::ShapeCastRequest*>(request),
-                s_sweepBuffer, m_pxScene, queryData, m_shapecastBufferSize);
-        }
-        else if (azrtti_istypeof<AzPhysics::OverlapRequest>(request))
+        switch (request->m_requestType)
         {
-            return Internal::OverlapQuery(azdynamic_cast<const AzPhysics::OverlapRequest*>(request),
-                s_overlapBuffer, m_pxScene, queryData, m_overlapBufferSize);
-        }
-        else
-        {
-            AZ_Warning("Physx", false, "Unknown Scene Query request type.");
-        }
-        return AzPhysics::SceneQueryHits();
+        case AzPhysics::SceneQueryRequest::Raycast:
```

Comment:
I'm surprised that this change is significantly faster. The code
```cpp
azrtti_istypeof<AzPhysics::RayCastRequest>(request)
```
is a macro that expands to
```cpp
AZ::RttiIsTypeOf<AzPhysics::RayCastRequest>(request))
```

Substituting in our types in this case, that template function is calling:
```cpp
AZ::Internal::RttiIsTypeOfHelper<AzPhysics::RayCastRequest, AzPhysics::SceneQueryRequest*>::Check(request, typename AZ::HasAZRtti<AzPhysics::SceneQueryRequest>::kind_type())
```

_That_ function is the same as
```cpp
request && request->RTTI_IsTypeOf(AZ::AzTypeInfo<AzPhysics::RayCastRequest>::Uuid()))
```

Now, that `RTTI_IsTypeOf` function is virtual. For the two classes in question, `RayCastRequest` and `SceneQueryRequest`, they have a `IsTypeOf` overload as:
```cpp
virtual inline bool RTTI_IsTypeOf(const AZ ::TypeId& typeId) const
{
    return RTTI_IsContainType(typeId);
}
```

Continuing down the rabbit hole, we get to the actual test:
```cpp
static constexpr bool SceneQueryRequest::RTTI_IsContainType(const AZ ::TypeId& id)
{
    return id == RTTI_Type();
}

static bool RayCastRequest::RTTI_IsContainType(const AZ ::TypeId& id)
{
    if (id == RTTI_Type())
    {
        return true;
    }
    return AZ ::Internal ::RttiCaller<SceneQueryRequest>::RTTI_IsContainType(id);
}
```

Those `id == RTTI_Type()` calls is doing the real comparison:
```cpp
constexpr bool Uuid::operator==(const Uuid& rhs) const
{
    return AZStd::ranges::equal(m_data, rhs.m_data);
}
```
which compares the `AZStd::byte[16]`'s of each Uuid.

From what I can tell, there's 1 virtual function call, and 2 `AZStd::byte[16]` comparisons for each call to `azrtti_istypeof`. So where's the speed going?

Reply:
Instead of performing an `azrtti_istypeof` followed by an `azdynamic_cast`, it is quicker to just perform the `azdynamic_cast`
The `azrtti_istypeof` check is unnecessary when a cast is going to be performed as it is checking the same information twice.

Reply:
@burelc-amzn it's because of `AZ::Uuid::CreateStringSkipWarning` 
Similar issue to this: https://github.com/o3de/o3de/issues/12339
Up to 20% of the entire raycast call was RTTI.. 

I'll try test with `azdynamic_cast`, thanks @lumberyard-employee-dm

Reply:
Avoiding a redundant azrtti_istypeof is better and can be done anyways, but I expect you'll still see that azdynamic_cast is still expensive due to the RTTI functions. Ultimately, getting to the bottom of why we end up creating a Uuid from the string from the RTTI macro at runtime instead of having the Uuid itself created at compile time would be the biggest win across all systems.

Reply:
Actually it has to be created at runtime because RTTI_GetType() is a virtual function and therefore relies on a vtable created at runtime.

Reply:
Yes, still getting `CreateStringSkipWarnings` with `azdynamic_cast` too. 

```
AZ::Uuid::CreateStringSkipWarnings(AZStd::basic_string_view<char,AZStd::char_traits<char>> uuidString, bool skipWarnings) Line 94	C++
AZ::Uuid::CreateString(AZStd::basic_string_view<char,AZStd::char_traits<char>> uuidString) Line 84	C++
AZ::Uuid::Uuid(AZStd::basic_string_view<char,AZStd::char_traits<char>> uuidString) Line 60	C++
AzPhysics::RayCastRequest::TYPEINFO_Uuid() Line 166	C++
AzPhysics::RayCastRequest::RTTI_Type() Line 166	C++
AzPhysics::RayCastRequest::RTTI_AddressOf(const AZ::Uuid & id) Line 166	C++
AZ::Internal::RttiCastHelper<AzPhysics::RayCastRequest const *,AzPhysics::SceneQueryRequest const *>(const AzPhysics::SceneQueryRequest * ptr, const std::integral_constant<enum AZ::RttiKind,1> & __formal) Line 614	C++
AZ::RttiCast<AzPhysics::RayCastRequest const *,AzPhysics::SceneQueryRequest const *>(const AzPhysics::SceneQueryRequest * ptr) Line 913	C++
PhysX::PhysXScene::QueryScene(const AzPhysics::SceneQueryRequest * request) Line 1003	C++

```

I tried to see if it's possible to cache the UUID in a static local variable inside TYPEINFO_Uuid() since that function will be unique per class type but it seems there are some issues with constexpr in this case. 

Here's my attempt:
```
    static constexpr AZ::TypeId TYPEINFO_Uuid() \
    { \
        static AZ::TypeId cachedUuid = AZ::TypeId::CreateNull(); \
        if (cachedUuid.IsNull()) \
            cachedUuid = AZ::TypeId(_ClassUuid) + AZ::Internal::AggregateTypes<__VA_ARGS__>::Uuid(); \
        return cachedUuid; \
    }

error C3615: constexpr function 'AZ::AttributeData<AZ::Script::Attributes::ExcludeFlags>::TYPEINFO_Uuid' cannot result in a constant expression
```

Reply:
I working on the AZ TypeInfo code right now.
This should be addressed soon.

FYI, a static local cannot be used inside of a constexpr function.
It makes the function unable to be evaluated at compile time.

Reply:
It should avoid the runtime calls to parse Uuid strings if you use a static constexpr variable instead, like in d59c241ec8945f0420bab6ccca48ce27529776e6. But it sounds like there’s a larger change coming.